### ISSUE TITLE
Material Design theme

### ISSUE THREAD

**Issue by dxbykov** [DevExpress/DevExtreme#3558]
# The Problem
Material Design is getting popular and becoming one of the modern web UI standards. For web developers, it’s frequently a requirement for new apps. Implementing it from scratch is a tremendous task.
# The Solution
We decided to provide the Material Design theme for DevExtreme components out-of-the-box. You can enable it using the [predefined themes approach](https://js.devexpress.com/Documentation/18_1/Guide/Themes/Predefined_Themes/). 

The new theme is available as `dx.material.blue.light.css` in the DevEextreme npm, Bower and NuGet pre-release packages. Also, you can use npm CDN to reference it:

```html
  <link rel="stylesheet" href="https://unpkg.com/devextreme@18.1-unstable/dist/css/dx.common.css" />
  <link rel="stylesheet" href="https://unpkg.com/devextreme@18.1-unstable/dist/css/dx.material.blue.light.css" />
```

For Angular apps, follow the [corresponding guide](https://github.com/DevExpress/devextreme-angular#add-to-existing-app) to setup your app.

### DataGrid
![material-theme-grid](https://user-images.githubusercontent.com/8307562/38367715-20a37ee0-38ec-11e8-8495-65e6075c292f.png)

### TreeList
![material-theme-treelist](https://user-images.githubusercontent.com/8307562/38367722-236c25e6-38ec-11e8-8252-355ff17b9580.png)

### Form
![material-theme-form](https://user-images.githubusercontent.com/8307562/38367729-26339502-38ec-11e8-8443-fbeb1e04dc0f.png)

# Try It And Give Us Feedback 

## Live Sandbox

[jQuery](https://codepen.io/anon/pen/LdBxqe?editors=1010) 
[Angular](https://plnkr.co/edit/DxEJAiJfO9P6AT5S8rAi?p=preview)
[React](https://stackblitz.com/edit/devextreme-react)
[Vue](https://embed.plnkr.co/Jqnd4HB7CpFp1Xd3Ngj3/)

## Installation

Follow the installation guide from the [release page](https://github.com/DevExpress/DevExtreme/releases) and leave us feedback on this topic.

**Comment by brux88** [comment]
hi it is possible change primary and accent color?

**Comment by babich-a** [comment]
@brux88, we are working on adding the Material theme to the [ThemeBuilder](https://js.devexpress.com/ThemeBuilder/). Also, we plan to release a number of predefined color schemes for the Material theme. You can now change the @base-accent variable in [material.blue.light.less](https://github.com/DevExpress/DevExtreme/blob/18_1/styles/widgets/material/color-schemes/blue.light/material.blue.light.less) and build your own dx.material.blue.light.css, following our [Developer's Guide](https://github.com/DevExpress/DevExtreme/blob/18_1/README_DEVELOPERS.md)

**Comment by rprimora** [comment]
Some widgets look like no work has been done on them since they do not fit in material theme:
- dxLookup
- dxSelectBox
- dxTextArea
- dxTextBox
- dxNumberBox

Context menu items have buttons? I think what android theme had for context menu is closer to material theme.

Also it would be awesome if you could provide an easy way of defining a template for no data. I am talking here about widgets that display data items, like grid or list. At the moment we can only define text that will be displayed when no data is available but I think providing us with a template function and giving us a abilitay to define empty template would be most beneficial. For instance, we could put an image of some sort, a text description and maybe a button to add new data in case no data is displayed.

Is toast message style pulled from some other theme at the moment or are you rolling with this. At the moment I get green toast bacground for success when the style should be pretty much the same as what you have in android theme.

Antoher thing I would like to add is the popup buttons. When dxButton is placed in popup via toolbar it gets ripped of its style and it becomes a flat button. It would be really nice if you could give us an option to retain the styling applied via "type" property(There are workarounds but they are ugly).

That being said I really like the new look and feel of grid. It is much more spacey and cleaner than that of android theme.

**Comment by groshenkovamarina** [comment]
Hey @rprimora! Thank you for your detailed feedback. We continue improving the Material theme and fixing some minor issues. That's why some of the widgets don't have a final look yet. For example, we have rolled out the dxToast with today's pre-release build and making some improvements in dxLookup right now.

Material guidelines don't describe every control and every feature we already have in DevExtreme. In some cases, we adapted the guidelines for our realities. For example, the difference between dxLookup and dxSelectBox is that we render SelectBox as a picker (we think it's ready) and Lookup as a dropdown menu (work on this hasn't been done yet). Here we follow [this part](https://material.io/guidelines/components/text-fields.html#text-fields-input-types) of guidelines. For the TextBox, TextArea and NumberBox we used [this part](https://material.io/guidelines/components/text-fields.html#text-fields-field-types) of guidelines. Right now, our editor controls don't have [a label option](https://material.io/guidelines/components/text-fields.html#text-fields-layout) so the only way to render editors with labels is to use the dxForm control. Right now, we don't have the focus animation of a label in a form but it's going to be added later this year. As for the label option in editors, we are discussing to add it in the next releases. Currently, we don't know about other issues here, please let us know if you see any differences with guidelines.

For the Toast control, please use the "info" value of type option to get the message with the gray background. For the other types, we use green, yellow and red colors from Material Design [color palette](https://material.io/guidelines/style/color.html#color-color-palette) to emphasize the differences of messages a user gets.

Right now, we use "dx-button-flat" CSS class to make a button flat and add use the same styles to all buttons inside toolbars and popups according to [Material Design guidelines](https://material.io/guidelines/components/dialogs.html). It's impossible to set any option to make a button 'non-flat' there right now. The good news is we are considering adding this option later this year for all available themes.

Talking about a collection widget template for 'no data', we added this idea to the internal product backlog for future discussion.

Not sure I got your question about the context menu. Can you describe the result you're expecting? A simple code block, an image or a guideline link would be helpful.

**Comment by sccrgoalie1** [comment]
> As for the label option in editors, we are discussing to add it in the next releases

Great.  This is a must have to make the editors look seamless in a material site.  The dxForm options works in some cases but won't work everywhere.

**Comment by rprimora** [comment]
@groshenkovamarina Here are some images of how it looks on my side.

This is dxLookup. It has a border all around which I think is not part of material look.
![dxlookup](https://user-images.githubusercontent.com/19466336/38886083-2ef8db24-4275-11e8-9a59-4ac834815775.jpg)
Thi is dxTextArea. It also has borders all around. I think this might be what you are saying - still in development.
![dxtextarea](https://user-images.githubusercontent.com/19466336/38886105-42aa317c-4275-11e8-8ec5-e4b091f39d99.jpg)

This is the hover effect of simple dxTextBox. The first input is the one I am hovering. In my opinion, the black border does not go well with this.
![hovereffect](https://user-images.githubusercontent.com/19466336/38886124-59062732-4275-11e8-85f2-a56fa6e9e38f.jpg)

Forgive me for mentioning dxContext menu, It seems I have implemented dxActionSheet instead of context menu.

I know this is still work in development and I am really looking forward to this theme but at this point android theme is much closer to material guidlines than this them.

**Comment by BingoRUS** [comment]
@rprimora Thanks for the clarification.
Yes, the dxLookup doesn't have a Material Design look yet. I'll let you know when we merge fixes of this control.
 
TextArea works as expected - check [this image](https://storage.googleapis.com/material-design/publish/material_v_12/assets/0B5ZSepuCX1xOYXNnRjJKVU53Skk/area-states.png) from Material Design guidelines. Opposite of the TextArea with a fixed height, the multi-line editor has a floating height and only a bottom border but we don't support multi-line editing for the dxTextBox. Do you need this kind of component in your application?
 
The hover effect of dxTextBox also works according to the guidelines - please check [the image with different states](https://storage.googleapis.com/material-design/publish/material_v_12/assets/0B5ZSepuCX1xOb195LU9KWEdIeEk/states.png) from [this page](https://material.io/guidelines/components/text-fields.html#text-fields-states).

**Comment by rprimora** [comment]
I gotta admit I did not scroll down to the textArea explanation but having seen it now I completely despise it 😄 

Yes I was thinking more along the lines of multi-line editor like so...
![ezgif-4-ed8328ad97](https://user-images.githubusercontent.com/19466336/38888669-af7fd9d0-427c-11e8-8b4d-1dee933b47a7.gif)
...this currently works in your android theme for dxTextArea and it can be seen in some places on web for instance when writing YouTube comments. It would be nice to support this so we are not limited by dxTextArea when wanting to make a multiline input.

I can see now the hover effect on the google material guidlines and it kinda looks different than your but I checked and the color is indeed black so I don't know. Might be my screen.

Also, do you have plans on implementing character counter below input border?

**Comment by BingoRUS** [comment]
@rprimora I've added the multi-line mode and the character counter to our product backlog for future discussion. And I would like to say thank you once again for your detailed feedback.

**Comment by pfeigl** [comment]
Lookup dropdown in DataTable is too small in material theme. See the following two screenshots of android vs material theme:

Good (android theme)
![image](https://user-images.githubusercontent.com/237891/40222725-c2ccce00-5a80-11e8-942e-16c367e35e37.png)

Not working (material theme)
![image](https://user-images.githubusercontent.com/237891/40222737-cf6d02ba-5a80-11e8-96d1-26ae3a102934.png)

**Comment by groshenkovamarina** [comment]
Hi @pfeigl!
The drop-down window width is equal to the editor width by default. This logic doesn't depend on a particular theme. 
At the same time, we added default paddings for each column in the material theme (see our [blog post](https://community.devexpress.com/blogs/donw/archive/2018/04/26/devextreme-new-material-design-theme-v18-1.aspx) for more information).
So, the editor width is small and is not enough to display text. We recommend you extend this column.
We appreciate you pointing out to this behavior and will consider improving it in the future.

**Comment by pfeigl** [comment]
Already realised that the lookup extends in size as soon as the column gets bigger.

What I don't understand is: Why is the dropdown so much bigger than the original field with the android theme? There is is also much bigger than the actual column (see my first screenshot)

Anyways: Big thanks for the new theme, it's looking fantastic, very nice work there!

**Comment by brux88** [comment]
hi, I have the same problem, because the descent was like the column, before it was bigger, how can I fix it

**Comment by groshenkovamarina** [comment]
@pfeigl Thank you for your positive feedback! You're right, in the android theme the width of the drop-down window is calculated differently. We'll check if we can move the same logic to the material theme.

**Comment by groshenkovamarina** [comment]
Hi @brux88,
As I wrote above, in the material theme default paddings of the grid column were increased. 
If extending the column does not suit you, then you can reduce the paddings using CSS. 
If your scenario is more complex, I recommend you contact our Support Team in the [Support Center](https://www.devexpress.com/Support/Center/Question/Create).

**Comment by BingoRUS** [comment]
Thanks for trying the feature. Based on your feedback, we improved some use cases in the Material theme. It is available since v18.1.3. Also, we extended a list of planned features for the next major release. Right now, I'm closing this discussion thread and recommend you create [a new issue here](https://github.com/DevExpress/DevExtreme/issues/new) or [a ticket in our Support Center](https://www.devexpress.com/Support/Center/Question/Create) in case of any bug, question or feature request related to Material Design and DevExtreme widgets.
